File: ios/Classes/PianoAnalyticsPlugin.swift
import Flutter
import UIKit
import PianoAnalytics

public class PianoAnalyticsPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "piano_analytics_plugin", binaryMessenger: registrar.messenger())
        let instance = PianoAnalyticsPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch (call.method) {
        case PAEvents.setConfiguration:
            
            guard   let args = call.arguments as? Dictionary<String, AnyObject>,
                    let collectDomain = args["collectDomain"] as? String,
                    let site = args["site"] as? Int
            else{
                result(FlutterError(code: "500", message: "Error setConfiguration :: collectDomain and site parameters is required", details: nil))
                return
            }
            
            let path: String? = args["path"] as? String? ?? nil
            let customUserAgent: String? = args["customUserAgent"] as? String? ?? nil
            
            let crashDetection: Bool = args["crashDetection"] as? Bool ?? PADefaultConfiguration.DEFAULT_CRASH_DETECTION
            
            let sessionBackgroundDuration: Int = args["sessionBackgroundDuration"] as? Int ?? PADefaultConfiguration.DEFAULT_SESSION_BACKGROUND_DURATION
            
            let ignoreLimitedAdvertisingTracking: Bool = args["ignoreLimitedAdvertisingTracking"] as? Bool ?? PADefaultConfiguration.DEFAULT_IGNORE_LIMITED_ADVERTISING_TRACKING
            
            let sendEventWhenOptout: Bool = args["sendEventWhenOptout"] as? Bool ?? PADefaultConfiguration.DEFAULT_SEND_WHEN_OPT_OUT
            
            let storageLifetimePrivacy: Int = args["storageLifetimePrivacy"] as? Int ?? PADefaultConfiguration.DEFAULT_STORAGE_LIFETIME_PRIVACY
            let storageLifetimeUser: Int = args["storageLifetimeUser"] as? Int ?? PADefaultConfiguration.DEFAULT_STORAGE_LIFETIME_USER
            let storageLifetimeVisitor: Int = args["storageLifetimeVisitor"] as? Int ?? PADefaultConfiguration.DEFAULT_STORAGE_LIFETIME_VISITOR
            
            let visitorId: String? = args["visitorId"] as? String? ?? nil
            
            var privacyDefaultModeP = PADefaultConfiguration.PRIVACY_DEFAULT_MODE
            if let privacyDefaultModePString: String = args["privacyDefaultMode"] as? String {
                privacyDefaultModeP = PrivacyDefaultModeP(rawValue: privacyDefaultModePString) ?? PADefaultConfiguration.PRIVACY_DEFAULT_MODE
            }
            
            var offlineEncryptionModeP = PADefaultConfiguration.DEFAULT_ENCRYPTION_MODE
            if let offlineEncryptionModePString: String = args["offlineEncryptionMode"] as? String {
                offlineEncryptionModeP = OfflineEncryptionModeP(rawValue: offlineEncryptionModePString) ?? PADefaultConfiguration.DEFAULT_ENCRYPTION_MODE
            }
            
            var offlineStorageModeP = PADefaultConfiguration.DEFAULT_OFFLINE_STORAGE_MODE
            if let offlineStorageModePString: String = args["offlineStorageMode"] as? String {
                offlineStorageModeP = OfflineStorageModeP(rawValue: offlineStorageModePString) ?? PADefaultConfiguration.DEFAULT_OFFLINE_STORAGE_MODE
            }
            
            var visitorStorageModeP = PADefaultConfiguration.DEFAULT_VISITOR_STORAGE_MODE
            if let visitorStorageModePString: String = args["visitorStorageMode"] as? String {
                visitorStorageModeP = VisitorStorageModeP(rawValue: visitorStorageModePString) ?? PADefaultConfiguration.DEFAULT_VISITOR_STORAGE_MODE
            }
            
            var visitorIdTypeModeP = PADefaultConfiguration.DEFAULT_VISITOR_ID_TYPE_MODE
            if let visitorIdTypeModePString: String = args["visitorIdTypeMode"] as? String {
                visitorIdTypeModeP = VisitorIdTypeModeP(rawValue: visitorIdTypeModePString) ?? PADefaultConfiguration.DEFAULT_VISITOR_ID_TYPE_MODE
            }
            
            var config : ConfigurationBuilder = ConfigurationBuilder().withCollectDomain(collectDomain)
                .withSite(site)
                .enableCrashDetection(crashDetection)
                .withSessionBackgroundDuration(sessionBackgroundDuration)
                .enableIgnoreLimitedAdTracking(ignoreLimitedAdvertisingTracking)
                .enableSendEventWhenOptout(sendEventWhenOptout)
                .withPrivacyDefaultMode(privacyDefaultModeP.sdkType())
                .withOfflineEncryptionMode(offlineEncryptionModeP.sdkType().rawValue)
                .withOfflineStorageMode(offlineStorageModeP.sdkType().rawValue)
                .withStorageLifetimePrivacy(storageLifetimePrivacy)
                .withStorageLifetimeUser(storageLifetimeUser)
                .withStorageLifetimeVisitor(storageLifetimeVisitor)
                .withVisitorStorageMode(visitorStorageModeP.sdkType().rawValue)
                .withVisitorIdType(visitorIdTypeModeP.sdkType().rawValue)
            
            if let pathValue = path{
                config = config.withPath(pathValue)
            }
            
            if let customUserAgentValue = customUserAgent {
                config = config.withCustomUserAgent(customUserAgentValue)
            }
            
            if let visitorIdValue = visitorId {
                config = config.withVisitorIdType(visitorIdValue)
            }
            
            pa.setConfiguration(
                config
                    .build()
            )
            
            break
        case PAEvents.sendEvent:
            guard   let args = call.arguments as? Dictionary<String, AnyObject>,
                    let eventName = args["eventName"] as? String
            else{
                result(FlutterError(code: "500", message: "Error sendEvent :: eventName is required", details: nil))
                return
            }
            
            
            if let data = args["data"] as? Dictionary<String, AnyObject> {
                pa.sendEvent(Event(eventName, data: data))
            } else {
                pa.sendEvent(Event(eventName, data: [:]))
            }
            
            break
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/com/plugin/piano_analytics_plugin/PianoAnalyticsPlugin.kt
package com.plugin.piano_analytics_plugin

import androidx.annotation.NonNull
import android.util.Log

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import io.piano.analytics.Event;
import io.piano.analytics.PianoAnalytics;
import io.piano.analytics.Configuration;
import io.piano.analytics.Configuration.VisitorIDType;
import io.piano.analytics.Configuration.VisitorStorageMode;
import io.piano.analytics.Configuration.OfflineStorageMode;
import io.piano.analytics.Configuration.EncryptionMode;

/** PianoAnalyticsPlugin */
class PianoAnalyticsPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var pa: PianoAnalytics

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "piano_analytics_plugin")
        channel.setMethodCallHandler(this)
        pa = PianoAnalytics.getInstance(flutterPluginBinding.applicationContext)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        try {

            val arguments: HashMap<String, Any> = call.arguments as HashMap<String, Any>

            when (call.method) {
                PAEvents.SET_CONFIGURATION -> {
                    val collectDomain: String = arguments["collectDomain"] as String

                    /// Return error if collectDomain is empty
                    if (collectDomain.isEmpty()) {
                        Log.e("PianoAnalyticsPlugin", "collectDomain is required")
                        result.error("500", "collectDomain is required", null)
                        return
                    }

                    val site: Int = arguments["site"] as Int
                    val path: String? = arguments["path"] as String?
                    val customUserAgent: String? = arguments["customUserAgent"] as String?
                    val crashDetection: Boolean = arguments["crashDetection"] as Boolean ?: PADefaultConfiguration.DEFAULT_CRASH_DETECTION
                    val sessionBackgroundDuration: Int = arguments["sessionBackgroundDuration"] as Int ?: PADefaultConfiguration.DEFAULT_SESSION_BACKGROUND_DURATION
                    val ignoreLimitedAdvertisingTracking: Boolean = arguments["ignoreLimitedAdvertisingTracking"] as Boolean ?: PADefaultConfiguration.DEFAULT_IGNORE_LIMITED_ADVERTISING_TRACKING
                    val sendEventWhenOptout: Boolean = arguments["sendEventWhenOptout"] as Boolean ?: PADefaultConfiguration.DEFAULT_SEND_WHEN_OPT_OUT

                    var privacyDefaultModeP: PrivacyDefaultModeP
                    try {
                        privacyDefaultModeP = PrivacyDefaultModeP.valueOf(arguments["privacyDefaultMode"] as String)
                    } catch (e: Exception) {
                        privacyDefaultModeP = PrivacyDefaultModeP.valueOf(PADefaultConfiguration.PRIVACY_DEFAULT_MODE)
                    }

                    var offlineEncryptionModeP: OfflineEncryptionModeP
                    try {
                        offlineEncryptionModeP = OfflineEncryptionModeP.valueOf(arguments["offlineEncryptionMode"] as String)
                    } catch (e: Exception) {
                        offlineEncryptionModeP = OfflineEncryptionModeP.valueOf(PADefaultConfiguration.DEFAULT_ENCRYPTION_MODE)
                    }

                    var offlineStorageModeP: OfflineStorageModeP
                    try {
                        offlineStorageModeP = OfflineStorageModeP.valueOf(arguments["offlineStorageMode"] as String)
                    } catch (e: Exception) {
                        offlineStorageModeP = OfflineStorageModeP.valueOf(PADefaultConfiguration.DEFAULT_OFFLINE_STORAGE_MODE)
                    }

                    val storageLifetimePrivacy: Int = arguments["storageLifetimePrivacy"] as Int ?: PADefaultConfiguration.DEFAULT_STORAGE_LIFETIME_PRIVACY
                    val storageLifetimeUser: Int = arguments["storageLifetimeUser"] as Int ?: PADefaultConfiguration.DEFAULT_STORAGE_LIFETIME_USER
                    val storageLifetimeVisitor: Int = arguments["storageLifetimeVisitor"] as Int ?: PADefaultConfiguration.DEFAULT_STORAGE_LIFETIME_VISITOR

                    var visitorStorageModeP: VisitorStorageModeP = VisitorStorageModeP.fixed
                    try {
                        visitorStorageModeP = VisitorStorageModeP.valueOf(arguments["visitorStorageMode"] as String)
                    } catch (e: Exception) {
                        visitorStorageModeP = VisitorStorageModeP.valueOf(PADefaultConfiguration.DEFAULT_VISITOR_STORAGE_MODE)
                    }

                    var visitorIdTypeModeP: VisitorIdTypeModeP = VisitorIdTypeModeP.uuid
                    try {
                        visitorIdTypeModeP = VisitorIdTypeModeP.valueOf(arguments["visitorIdTypeMode"] as String)
                    } catch (e: Exception) {
                        visitorIdTypeModeP = VisitorIdTypeModeP.valueOf(PADefaultConfiguration.DEFAULT_VISITOR_ID_TYPE_MODE)
                    }

                    val visitorId: String? = arguments["visitorId"] as String?

                    pa.setConfiguration(
                        Configuration.Builder()
                            .withCollectDomain(collectDomain)
                            .withSite(site)
                            .withPath(path)
                            .withCustomUserAgent(customUserAgent)
                            .enableCrashDetection(crashDetection)
                            .withSessionBackgroundDuration(sessionBackgroundDuration)
                            .enableIgnoreLimitedAdTracking(ignoreLimitedAdvertisingTracking)
                            .enableSendEventWhenOptOut(sendEventWhenOptout)
                            .withPrivacyDefaultMode(privacyDefaultModeP.sdkName)
                            .withEncryptionMode(offlineEncryptionModeP.sdkName)
                            .withOfflineStorageMode(offlineStorageModeP.sdkName)
                            .withStorageLifetimePrivacy(storageLifetimePrivacy)
                            .withStorageLifetimeUser(storageLifetimeUser)
                            .withStorageLifetimeVisitor(storageLifetimeVisitor)
                            .withVisitorStorageMode(visitorStorageModeP.sdkName)
                            .withVisitorIDType(visitorIdTypeModeP.sdkName)
                            .withVisitorID(visitorId)
                            .build()
                    )
                }

                PAEvents.SEND_EVENT -> {
                    val eventName: String = arguments["eventName"] as String

                    if (eventName.isEmpty()) {
                        result.error("500", "eventName is required", null)
                    } else {
                        val data: HashMap<String, Any?> = arguments["data"] as HashMap<String, Any?>
                        pa.sendEvent(Event(eventName, data))
                    }
                }

                else -> result.notImplemented()
            }
        } catch (e: Exception) {
            Log.e("PianoAnalyticsPlugin", e.toString())
            result.error("500", e.toString(), null)
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}
